src/aquarium-optimized/Aquarium.cpp

```diff
@@ -786,219 +782,112 @@ void Aquarium::render() {
-  matrix::mulMatrixMatrix4(worldUniforms.worldViewProjection,
-                           worldUniforms.world,
-                           lightWorldPositionUniform.viewProjection);
-  matrix::inverse4(g.worldInverse, worldUniforms.world);
-  matrix::transpose4(worldUniforms.worldInverseTranspose, g.worldInverse);
-}
+        model->updatePerInstanceUniforms(worldUniforms);
 
-void Aquarium::updateWorldMatrixAndDraw(Model *model) {
-  if (model->worldmatrices.size()) {
-    for (auto &world : model->worldmatrices) {
-      updateWorldProjections(world);
+        if (!drawPerModel) {
+          model->draw();
+        }
+      }
+    } else {
+      FishModel *model = static_cast<FishModel *>(mAquariumModels[i]);
       model->prepareForDraw();
-      model->updatePerInstanceUniforms(worldUniforms);
-      model->draw();
+
+      const Fish &fishInfo = fishTable[i - fishBegin];
+      int numFish = fishCount[i - fishBegin];
+      float fishBaseClock = g.mclock * g_fishSpeed;
+      float fishRadius = fishInfo.radius;
+      float fishRadiusRange = fishInfo.radiusRange;
+      float fishSpeed = fishInfo.speed;
+      float fishSpeedRange = fishInfo.speedRange;
+      float fishTailSpeed = fishInfo.tailSpeed * g_fishTailSpeed;
+      float fishOffset = g_fishOffset;
+      // float fishClockSpeed  = g_fishSpeed;
+      float fishHeight = g_fishHeight + fishInfo.heightOffset;
+      float fishHeightRange = g_fishHeightRange * fishInfo.heightRange;
+      float fishXClock = g_fishXClock;
+      float fishYClock = g_fishYClock;
+      float fishZClock = g_fishZClock;
+
+      for (int ii = 0; ii < numFish; ++ii) {
+        float fishClock = fishBaseClock + ii * fishOffset;
+        float speed = fishSpeed + static_cast<float>(matrix::pseudoRandom()) *
+                                      fishSpeedRange;
+        float scale = 1.0f + static_cast<float>(matrix::pseudoRandom()) * 1;
+        float xRadius =
+            fishRadius +
+            static_cast<float>(matrix::pseudoRandom()) * fishRadiusRange;
+        float yRadius =
+            2.0f + static_cast<float>(matrix::pseudoRandom()) * fishHeightRange;
+        float zRadius =
+            fishRadius +
+            static_cast<float>(matrix::pseudoRandom()) * fishRadiusRange;
+        float fishSpeedClock = fishClock * speed;
+        float xClock = fishSpeedClock * fishXClock;
+        float yClock = fishSpeedClock * fishYClock;
+        float zClock = fishSpeedClock * fishZClock;
+
+        model->updateFishPerUniforms(
+            sin(xClock) * xRadius, sin(yClock) * yRadius + fishHeight,
+            cos(zClock) * zRadius, sin(xClock - 0.04f) * xRadius,
+            sin(yClock - 0.01f) * yRadius + fishHeight,
+            cos(zClock - 0.04f) * zRadius, scale,
+            fmod((g.mclock + ii * g_tailOffsetMult) * fishTailSpeed * speed,
+                 static_cast<float>(M_PI) * 2),
+            ii);
+
+        model->updatePerInstanceUniforms(worldUniforms);
```

Comment:
Should this also be guarded by `!drawPerModel`?

Reply:
done